x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1641,7 +1645,29 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
                 sslConfiguration = null;
                 populateClientCertificate = (channel, threadContext) -> {};
             }
-            return new Netty4HttpServerTransport(
+            final AuthenticationService authenticationService = this.authcService.get();
+            final ThreadContext threadContext = this.threadContext.get();
+            final Supplier<Netty4HttpHeaderValidator> authenticateHttpRequest = () -> HttpHeadersUtils.getValidatorInboundHandler(
+                (httpRequest, channel, listener) -> {
+                    if (threadContext.isDefaultContext() == false) {
+                        listener.onFailure(new IllegalStateException("HTTP authentication will not run on non-empty thread context"));
+                        return;
+                    }
+                    // step 1: Populate the thread context with credentials and any other HTTP request header values (eg run-as) that the
+                    // authentication process looks for while doing its duty.
+                    perRequestThreadContext.accept(httpRequest, threadContext);
+                    populateClientCertificate.accept(channel, threadContext);
+                    RemoteHostHeader.process(channel, threadContext);
+                    // step 2: Run authentication on the now properly prepared thread-context.
+                    // This inspects and modifies the thread context.
+                    authenticationService.authenticate(
+                        httpRequest,
+                        ActionListener.wrap(ignored -> listener.onResponse(null), listener::onFailure)
+                    );
+                },
+                threadContext
+            );
```

Comment:
As part of the progress to simplify the logic of the authenticator here, I've raised https://github.com/elastic/elasticsearch/pull/95498 .